From c8544612932d8d2d6ebe78ed93c3c3dc7c7c33f8 Mon Sep 17 00:00:00 2001
From: Radu Bulie <radu.bulie@freescale.com>
Date: Thu, 18 Sep 2014 14:53:21 +0000
Subject: [PATCH 817/987] dpa_offload: Add check for initialization vector

This patch adds a check for the maximum value of the IV
provided to an outbound SA.

Change-Id: I102aeb893100222171103a738f0bfa20ed17f33c
Signed-off-by: Radu Bulie <radu.bulie@freescale.com>
Reviewed-on: http://git.am.freescale.net:8181/19121
Reviewed-by: Marian-Cornel Chereji <marian.chereji@freescale.com>
Tested-by: Marian-Cornel Chereji <marian.chereji@freescale.com>
[Lu:Original patch taken from
QorIQ-SDK-V1.7-SOURCE-20141218-yocto.iso]
Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
---
 drivers/staging/fsl_dpa_offload/wrp_dpa_ipsec.c | 9 +++++++++
 include/linux/fsl_dpa_ipsec.h                   | 1 +
 2 files changed, 10 insertions(+)

diff --git a/drivers/staging/fsl_dpa_offload/wrp_dpa_ipsec.c b/drivers/staging/fsl_dpa_offload/wrp_dpa_ipsec.c
index 34747d6..ddff1ba 100644
--- a/drivers/staging/fsl_dpa_offload/wrp_dpa_ipsec.c
+++ b/drivers/staging/fsl_dpa_offload/wrp_dpa_ipsec.c
@@ -411,6 +411,15 @@ static int do_copy_sa_params(struct dpa_ipsec_sa_params *prm, void *args)
 				kfree(sa_out_iv);
 				return err;
 			}
+
+			if (sa_out_iv->length > DPA_IPSEC_MAX_IV_LEN) {
+				err = -EINVAL;
+				log_err("Error - IV length greater than %d\n",
+					DPA_IPSEC_MAX_IV_LEN);
+				kfree(sa_out_iv);
+				return err;
+			}
+
 			sa_out_prm->init_vector = sa_out_iv;
 
 			/* if the IV array is NULL, don't bother to copy it */
diff --git a/include/linux/fsl_dpa_ipsec.h b/include/linux/fsl_dpa_ipsec.h
index 41e22d3..63e8601 100644
--- a/include/linux/fsl_dpa_ipsec.h
+++ b/include/linux/fsl_dpa_ipsec.h
@@ -73,6 +73,7 @@
 		2 * PORT_FIELD_LEN)
 
 
+#define DPA_IPSEC_MAX_IV_LEN         16   /* Maximum length of IV(in bytes) */
 #define DPA_IPSEC_MAX_POL_PER_SA     255  /* Maximum supported number of
 					   * policies per  SA              */
 
-- 
1.9.1

